cache/disk/disk.go

```diff
@@ -288,161 +242,220 @@ func (c *Cache) Put(kind cache.EntryKind, hash string, expectedSize int64, r io.
 
 	key := cacheKey(kind, hash)
 
-	c.mu.Lock()
+	var tf *os.File // Tempfile.
 
-	// If there's an ongoing upload (i.e. cache key is present in uncommitted state),
-	// we drop the upload and discard the incoming stream. We do accept uploads
-	// of existing keys, as it should happen relatively rarely (e.g. race
-	// condition on the bazel side) but it's useful to overwrite poisoned items.
-	if existingItem, found := c.lru.Get(key); found {
-		if !existingItem.(*lruItem).committed {
-			c.mu.Unlock()
-			io.Copy(ioutil.Discard, r)
-			return nil
+	// Cleanup intermediate state if something went wrong and we
+	// did not successfully commit.
+	unreserve := false
+	removeTempfile := false
+	defer func() {
+		// No lock required to remove stray tempfiles.
+		if removeTempfile {
+			os.Remove(tf.Name())
 		}
-	}
 
-	// Try to add the item to the LRU.
-	newItem := &lruItem{
-		size:      expectedSize,
-		committed: false,
-	}
-	ok := c.lru.Add(key, newItem)
-	c.mu.Unlock()
-	if !ok {
-		return &cache.Error{
-			Code: http.StatusInsufficientStorage,
-			Text: fmt.Sprintf("The item is larger (%d) than the cache's maximum size (%d).",
-				expectedSize, c.lru.MaxSize()),
+		if unreserve {
+			c.mu.Lock()
+			err := c.lru.Unreserve(size)
+			if err != nil {
+				// Set named return value.
+				rErr = err
+				log.Printf(rErr.Error())
+			}
+			c.mu.Unlock()
 		}
-	}
+	}()
 
-	// By the time this function exits, we should either mark the LRU item as committed
-	// (if the upload went well), or delete it. Capturing the flag variable is not very nice,
-	// but this stuff is really easy to get wrong without defer().
-	shouldCommit := false
-	filePath := cacheFilePath(kind, c.dir, hash)
-	defer func() {
+	if size > 0 {
 		c.mu.Lock()
-		if shouldCommit {
-			newItem.committed = true
-		} else {
-			c.lru.Remove(key)
+		ok, err := c.lru.Reserve(size)
+		if err != nil {
+			c.mu.Unlock()
+			return &cache.Error{
+				Code: http.StatusInternalServerError,
+				Text: err.Error(),
+			}
 		}
-		c.mu.Unlock()
-
-		if shouldCommit && c.proxy != nil {
-			// TODO: buffer in memory, avoid a filesystem round-trip?
-			fr, err := os.Open(filePath)
-			if err == nil {
-				c.proxy.Put(kind, hash, expectedSize, fr)
+		if !ok {
+			c.mu.Unlock()
+			return &cache.Error{
+				Code: http.StatusInsufficientStorage,
+				Text: fmt.Sprintf("The item (%d) + reserved space is larger than the cache's maximum size (%d).",
+					size, c.lru.MaxSize()),
 			}
 		}
-	}()
+		c.mu.Unlock()
+		unreserve = true
+	}
+
+	// Final destination, if all goes well.
+	filePath := cacheFilePath(kind, c.dir, hash)
+
+	// We will download to this temporary file.
+	tf, err := tfc.Create(filePath)
+	if err != nil {
+		return err
+	}
+	removeTempfile = true
 
-	// Download to a temporary file
-	tmpFilePath := filePath + ".tmp"
-	f, err := os.Create(tmpFilePath)
+	err = writeAndCloseFile(r, kind, hash, size, tf)
 	if err != nil {
 		return err
 	}
+
+	if c.proxy != nil {
+		rc, err := os.Open(tf.Name())
```

Comment:
Smart that you use the temporary file, instead of the final destination file, for the proxy, so that no lock is needed.